server/src/main/java/org/elasticsearch/index/mapper/KeyedJsonAtomicFieldData.java

```diff
@@ -0,0 +1,230 @@
+            this.delegate = delegate;
+            this.minOrd = minOrd;
+            this.maxOrd = maxOrd;
+        }
+
+        @Override
+        public long getValueCount() {
+            return delegate.getValueCount();
+        }
+
+        /**
+         * Returns the (un-prefixed) term value for the requested ordinal.
+         *
+         * Note that this method can only be called on ordinals returned from {@link #nextOrd()}.
+         * Otherwise it may attempt to look up values that do not share the correct prefix, which
+         * can result in undefined behavior or an error.
+         */
+        @Override
+        public BytesRef lookupOrd(long ord) throws IOException {
+            BytesRef keyedValue = delegate.lookupOrd(ord);
+            int prefixLength = key.length + 1;
+            int valueLength = keyedValue.length - prefixLength;
+            return new BytesRef(keyedValue.bytes, prefixLength, valueLength);
+        }
+
+        @Override
+        public long nextOrd() throws IOException {
+            if (minOrd < 0) {
+                return NO_MORE_ORDS;
+            }
+
+            for (long ord = delegate.nextOrd(); ord != NO_MORE_ORDS; ord = delegate.nextOrd()) {
+                if (minOrd <= ord) {
+                    if (ord <= maxOrd) {
+                        return ord;
+                    } else {
+                        return NO_MORE_ORDS;
+                    }
+                }
+            }
+            return NO_MORE_ORDS;
+        }
+
+        @Override
+        public boolean advanceExact(int target) throws IOException {
+            if (minOrd < 0) {
+                return false;
+            }
+
+            if (delegate.advanceExact(target)) {
+                for (long ord = delegate.nextOrd(); ord != NO_MORE_ORDS; ord = delegate.nextOrd()) {
+                     if (minOrd <= ord && ord <= maxOrd) {
+                         boolean advanced = delegate.advanceExact(target);
+                         assert advanced;
+                         return true;
+                    }
+                }
+            }
+            return false;
+        }
+
+        /**
+         * Performs a binary search to find the first term with 'key' as a prefix.
+         */
+        static long findMinOrd(BytesRef key, SortedSetDocValues delegate) throws IOException {
+            long low = 0;
+            long high = delegate.getValueCount() - 1;
+
+            long result = -1;
+            while (low <= high) {
+                long mid = (low + high) / 2;
+                final BytesRef term = delegate.lookupOrd(mid);
+                int cmp = compare(key, term);
+
+                if (cmp == 0) {
+                    result = mid;
+                    high = mid - 1;
+                } else if (cmp < 0) {
+                    high = mid - 1;
+                } else {
+                    low = mid + 1;
+                }
+            }
+            return result;
+        }
```

Comment:
Doc values already have an optimized way to find the first term that has a prefix via their terms enum, so we could do something like that. In general the underlying implementation does a binary search as well, but often it is a bit more efficient eg. by working directly on the compressed data.

```java
TermsEnum te = sortedSetDocValues.termsEnum();
if (te.seekCeil(prefix) != SeekStatus.END && StringHelper.startsWith(te.term(), prefix)) {
  return te.ord();
} else {
  return -1;
}
```

Reply:
For `SortedSetDocValues` it looks like this will call into `SortedDocValues#lookupTerm`, which is very similar to the current implementation. I find the current approach cleaner in that `findMinOrd` and `findMaxOrd` follow the same template, and that comparisons happen only on the prefixes. Maybe I could keep it as is for now, but see if it makes a difference when I run some benchmarks (on the TODO list)?

Reply:
Fine with me.